### PR TITLE
reconstruct data reader and config

### PR DESCRIPTION
**Pull request by sunxl1988**
gpu: one P40
faster rcnn(batch size 1): 0.298 <-> 0.336 (previous) 
yolov3(batch size 8): 0.1883 <-> 0.201 (previous)

add prefetch for data loader
split config yml into 3 parts (reader, architecture, optimize)
add increment config method 

adapt mask rcnn & mask rcnn fpn

### CODE REVIEW

File: tools/train.py [hunk 1]
```diff
@@ -99,13 +98,21 @@ def run(FLAGS, cfg):
         random.seed(0)
         np.random.seed(0)
 
+    if cfg.use_gpu:
+        devices_num = fluid.core.get_cuda_device_count()
+    else:
+        devices_num = int(os.environ.get('CPU_NUM', 1))
```
Comment (jerrywgz) [inline tools/train.py:104]: actually devices_num is 1 here, multiprocess is used for imperative training

File: ppdet/data/sampler.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+        def _init_context():
+            communicator_prog = fluid.Program()
+            init_communicator(communicator_prog, strategy.local_rank,
+                              strategy.nranks, True, strategy.current_endpoint,
+                              strategy.trainer_endpoints)
+            exe = fluid.Executor(place)
+            exe.run(communicator_prog)
+
+        if fluid.in_dygraph_mode():
```
Comment (jerrywgz) [inline ppdet/data/sampler.py:175]: dygraph_mode branch is enough

File: ppdet/modeling/head/bbox_head.py [hunk 3]
```diff
@@ -5,10 +5,37 @@
...
-class TwoFCHead(Layer):
+class Res5Feat(Layer):
+    __shared__ = ['num_stages']
+
+    def __init__(self, feat_in=1024, feat_out=512, num_stages=1):
+        super(Res5Feat, self).__init__()
+        na = NameAdapter(self)
+        self.res5_conv = []
+        for i in range(num_stages):
+            if i == 0:
```
Comment (jerrywgz) [inline ppdet/modeling/head/bbox_head.py:21]: no need for this condition

File: tools/eval.py [hunk 4]
```diff
@@ -42,23 +41,24 @@ def parse_args():
 
 def run(FLAGS, cfg):
```
Comment (jerrywgz) [inline tools/eval.py:42]: place is needed here

File: ppdet/modeling/head/bbox_head.py [hunk 5]
```diff
@@ -132,11 +160,14 @@ def __init__(self,
     def forward(self, body_feats, rois, spatial_scale, stage=0):
         bbox_feat = self.bbox_feat(body_feats, rois, spatial_scale, stage)
         if self.with_pool:
-            bbox_feat = fluid.layers.pool2d(
+            bbox_feat_ = fluid.layers.pool2d(
                 bbox_feat, pool_type='avg', global_pooling=True)
+            bbox_feat_ = fluid.layers.squeeze(bbox_feat_, axes=[2, 3])
```
Comment (jerrywgz) [inline ppdet/modeling/head/bbox_head.py:165]: why squeeze here?

File: ppdet/data/transform/operators.py [hunk 6]
```diff
@@ -2363,109 +2450,6 @@ def __call__(self, sample, context=None):
         return sample
 
 
-@register_op
```
Comment (jerrywgz) [inline ppdet/data/transform/operators.py:2366]: just keep it

File: ppdet/data/transform/batch_operators.py [hunk 7]
```diff
@@ -1,17 +1,3 @@
-# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (jerrywgz) [inline ppdet/data/transform/batch_operators.py:1]: keep copyright here

File: configs/base/faster_rcnn_r50_1x.yml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+architecture: FasterRCNN
+pretrain_weights: https://paddlemodels.bj.bcebos.com/object_detection/dygraph/resnet50.pdparams
+weights: output/faster_rcnn_r50_1x/model_final
+use_gpu: true
+worker_num: 0
```
Comment (jerrywgz) [inline configs/base/faster_rcnn_r50_1x.yml:5]: worker_num should be defined in reader

File: configs/reader/faster_rcnn.yml [hunk 9]
```diff
@@ -2,10 +2,10 @@ TrainReader:
   inputs_def:
     fields: ['image', 'im_info', 'im_id', 'gt_bbox', 'gt_class', 'is_crowd']
   dataset:
-    !COCODataSet
+    !COCODataset
+    dataset_dir: /home/ai/dataset/COCO17/
```
Comment (jerrywgz) [inline configs/reader/faster_rcnn.yml:6]: all the path need to be corrected

File: configs/reader/faster_rcnn.yml [hunk 10]
```diff
@@ -26,22 +26,20 @@ TrainReader:
     channel_first: true
   batch_transforms:
   - !PadBatch
-    pad_to_stride: 0
+    pad_to_stride: 32
```
Comment (jerrywgz) [inline configs/reader/faster_rcnn.yml:29]: pad_to_stride is 0 in faster_rcnn

File: ppdet/data/__init__.py [hunk 11]
```diff
@@ -1,19 +1,4 @@
-# Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (jerrywgz) [inline ppdet/data/__init__.py:1]: keep the copyright

File: tools/train.py [hunk 12]
```diff
@@ -188,13 +186,13 @@ def main():
     merge_config(FLAGS.opt)
     check_config(cfg)
     check_gpu(cfg.use_gpu)
-    check_version()
+    #check_version()
```
[the remark this answers is not in the document]
Comment (sunxl1988) [reply tools/train.py:189]: 修改

File: ppdet/modeling/head/bbox_head.py (hunk 3 above)
Comment (sunxl1988) [reply ppdet/modeling/head/bbox_head.py:21]: 修改

File: tools/eval.py (hunk 4 above)
Comment (sunxl1988) [reply tools/eval.py:42]: 修改

File: ppdet/data/transform/batch_operators.py (hunk 7 above)
Comment (sunxl1988) [reply ppdet/data/transform/batch_operators.py:1]: 修改

**Pull request closed without merging**